src/hotspot/share/utilities/resourceHash.hpp

```diff
@@ -77,16 +78,22 @@ class ResourceHashtable : public ResourceObj {
 
   Node const** lookup_node(unsigned hash, K const& key) const {
     return const_cast<Node const**>(
-        const_cast<ResourceHashtable*>(this)->lookup_node(hash, key));
+        const_cast<ResourceHashtableBase*>(this)->lookup_node(hash, key));
   }
 
+ protected:
+  Node** table() const { return STORAGE::table(); }
+
  public:
-  ResourceHashtable() { memset(_table, 0, SIZE * sizeof(Node*)); }
+  ResourceHashtableBase() : _number_of_entries(0) {}
 
-  ~ResourceHashtable() {
-    if (ALLOC_TYPE == C_HEAP) {
-      Node* const* bucket = _table;
-      while (bucket < &_table[SIZE]) {
+  ResourceHashtableBase(unsigned size) : STORAGE(size), _number_of_entries(0) {}
+
+  ~ResourceHashtableBase() {
```

Comment:
Base class constructor should be non-public to avoid slicing.  Also need to decide what to do about copying, either disallow or deep copy.  Default shallow copy seems likely to be wrong.

Reply:
I made all the constructors and destructors protected. I also made the classes `NONCOPYABLE`.